## Dataset 1
Changing Opportunity: Sociological Mechanisms Underlying Growing Class Gaps and Shrinking Race Gaps in Economic Mobility

arquivo: county\_by\_cohort\_estimates

National-Level Outcomes by Birth Cohort, Parental Income, Race, and Gender

*Outcomes* de mobilidade social em nível nacional por coorte de nascimento (todos os nascidos em um determinado período, renda parental, raça, gênero)

---

Reports outcomes for children born between 1978 and 1992 at the national
level by race, gender, and parent percentile. There is exactly one row per birth cohort.

## Variáveis 

### cohort
coorte de nescimento - 1978 até 1992


### [outcome]\\_[race]\\_[gender]\\_p[pctile]
Média predita de *outcome* para nascidos de uma dada raça, gênero e pais dentro de um dado percentil na distribuição de renda familiar nacional.

[race] -> pode ser todas (pooled), branca (white), preta (black), hispânica (hisp), asiatica (asian) e indígena americano e nativo do Alaska (aian - American Indian and Alaskan Native)
[gender] -> pode ser ambos (pooled), masculino (male) ou feminino (female)
[pctile] -> pode ser 1, 25, 50, 75, 100

### [race]\\_pooled\\_count
Número de nascidos abaixo de 18 para cada coorte e raça, amostrados por gênero

### [race]\\_pooled\\_blw\\_p50\\_count
Nº de nascidos em cada coorte e raça, amostrados por gênero, com pais que ganham menos que a renda parental nacional mediana 

### outcomes
- kfr: classificação percentil média (relativa a nascimento no mesmo ano) na distribuição nacional de renda familiar (i.e., renda pessoal acrescida da renda do cônjuge) medida aos 27 anos.
- kir: classificação percentil média (relativa a nascimento no mesmo ano) na distribuição nacional de renda pessoal (renda própria) medida aos 27 anos.
- emp: fração de pessoas empregadas aos 27 anos.
- kfi: renda familiar média aos 27 anos (em $ de 2023) obtida convertendo classificações percenti (kfr) para dólares usando a conversão percentil-dólar
- kii: renda individual média aos 27 anos (em $ de 2023) obtida convertendo classificações percentil (kir) para dólares usando a conversão percentil-dólar

# Dataset 2
Reports changes in outcomes between children born in 1978 and 1992 by county, race, gender, and parent percentile. Each county is uniquely identified by two identifiers − state and county (2010 FIPS codes). There is one row per county in this file.

For each county, we provide estimates of changes in outcomes at the 1st, 25th, 50th, 75th, and 100th percentile of the parent national income distribution by race and gender. As mudanças na renda para cada conty, raça, gênero e renda parental são estimadas utilizando uma regressão linear da renda para coorte de nascimento e calculando a diferença entre as rendas previstas dessa regressão para nascidos em 1992 e nascidos em 1978. 

## Variáveis
- State - int - Two-digit state 2010 FIPS code
- state\_name - str - State name
- County - int - three-digit county 2010 FIPS code
- county\_name - str - county name
- change\_[outcome]\_[race]\_[gender]\_p[pctile] - num - Difference between [outcome]\_[race]\_[gender]\_p[pctile]\_1992 and [outcome]\_[race]\_[gender]\_p[pctile]\_1978

# Perguntas

1. Quais são as diferenças existentes entre os grupos?
   1. Homens ascendem mais que mulheres?
   2. Qual raça tem maior tendência de ascensão?
   3. Qual percentil de ganho dos pais tem maior chance de ascensão?
2. Houve alguma redução na desigualdade social/distâncias entre classes?

In [1]:
import pandas as pd
from pathlib import Path

In [2]:
DATA_DIR = Path("data")
RAW_DIR = DATA_DIR / "raw"

In [3]:
# Reading county by cohort estimates
df_county_cohort = pd.read_parquet(RAW_DIR / 'county_by_cohort_estimates.parquet')

In [4]:
df_county_trends_estimate = pd.read_parquet(RAW_DIR / "table_1_county_trends_estimates.parquet")

In [5]:
df_county_trends_estimate.head(10)

state  county state_name county_name  kfr_aian_female_p1_1978  \
0      1       1    Alabama     Autauga                      NaN   
1      1       3    Alabama     Baldwin                      NaN   
2      1       5    Alabama     Barbour                      NaN   
3      1       7    Alabama        Bibb                      NaN   
4      1       9    Alabama      Blount                      NaN   
5      1      11    Alabama     Bullock                      NaN   
6      1      13    Alabama      Butler                      NaN   
7      1      15    Alabama     Calhoun                      NaN   
8      1      17    Alabama    Chambers                      NaN   
9      1      19    Alabama    Cherokee                      NaN   

   kfr_aian_female_p1_1992  change_kfr_aian_female_p1  \
0                      NaN                        NaN   
1                      NaN                        NaN   
2                      NaN                        NaN   
3                      NaN                        NaN   
4                      NaN                        NaN   
5                      NaN                        NaN   
6                      NaN                        NaN   
7                      NaN                        NaN   
8                      NaN                        NaN   
9                      NaN                        NaN   

   change_kfr_aian_female_p1_se  kfr_aian_female_p1_reliab  \
0                           NaN                        NaN   
1                           NaN                        NaN   
2                           NaN                        NaN   
3                           NaN                        NaN   
4                           NaN                        NaN   
5                           NaN                        NaN   
6                           NaN                        NaN   
7                           NaN                        NaN   
8                           NaN                        NaN   
9                           NaN                        NaN   

   kfr_asian_female_p1_1978  ...  change_kfi_pooled_pooled_p100_se  \
0                       NaN  ...                         1866.7131   
1                       NaN  ...                         1244.0596   
2                       NaN  ...                         3079.3716   
3                       NaN  ...                         8064.7051   
4                       NaN  ...                         2403.4680   
5                       NaN  ...                         9601.2969   
6                       NaN  ...                         6396.2593   
7                       NaN  ...                         1614.5885   
8                       NaN  ...                         2946.7351   
9                       NaN  ...                         3949.8279   

   change_kii_pooled_pooled_p100_se  kfi_white_pooled_p100_1978  \
0                        1497.25980                   51098.230   
1                        1013.37040                   48504.352   
2                        1989.66610                   48021.953   
3                        5079.89260                   57029.688   
4                        2788.46660                   52930.375   
5                        8082.46530                   63972.246   
6                        4499.89360                   61187.004   
7                         815.10431                   50000.242   
8                        1808.49190                   51107.121   
9                        3176.31350                   57291.586   

   kii_white_pooled_p100_1978  kfi_white_pooled_p100_1992  \
0                   37512.977                   57154.227   
1                   38108.254                   51846.117   
2                   36410.723                   58466.305   
3                   42442.129                   54718.160   
4                   39683.320                   59671.676   
5                   47243.605                   51217.234   
6      

In [6]:
# Filtering only the change columns
filtered_df = df_county_trends_estimate.filter(like="change", axis=1)
filtered_df.head(10)

change_kfr_aian_female_p1  change_kfr_aian_female_p1_se  \
0                        NaN                           NaN   
1                        NaN                           NaN   
2                        NaN                           NaN   
3                        NaN                           NaN   
4                        NaN                           NaN   
5                        NaN                           NaN   
6                        NaN                           NaN   
7                        NaN                           NaN   
8                        NaN                           NaN   
9                        NaN                           NaN   

   change_kfr_asian_female_p1  change_kfr_asian_female_p1_se  \
0                         NaN                            NaN   
1                         NaN                            NaN   
2                         NaN                            NaN   
3                         NaN                            NaN   
4                         NaN                            NaN   
5                         NaN                            NaN   
6                         NaN                            NaN   
7                         NaN                            NaN   
8                         NaN                            NaN   
9                         NaN                            NaN   

   change_kfr_black_female_p1  change_kfr_black_female_p1_se  \
0                    0.007220                       0.048081   
1                    0.049207                       0.031633   
2                    0.007344                       0.041772   
3                    0.008007                       0.118000   
4                         NaN                            NaN   
5                    0.046668                       0.062700   
6                    0.010766                       0.024248   
7                    0.093872                       0.029341   
8                    0.033095                       0.026006   
9                         NaN                            NaN   

   change_kfr_hisp_female_p1  change_kfr_hisp_female_p1_se  \
0                        NaN                           NaN   
1                  -0.057139                      0.124691   
2                        NaN                           NaN   
3                        NaN                           NaN   
4                  -0.034755                      0.185880   
5                        NaN                           NaN   
6                        NaN                           NaN   
7                   0.184147                      0.078651   
8                        NaN                           NaN   
9                        NaN                           NaN   

   change_kfr_pooled_female_p1  change_kfr_pooled_female_p1_se  ...  \
0                    -0.060184                        0.033494  ...   
1                     0.020877                        0.022728  ...   
2                     0.004866                        0.034296  ...   
3                    -0.043698                        0.043633  ...   
4                    -0.032124                        0.041569  ...   
5                     0.035805                        0.064309  ...   
6                     0.055381                        0.029481  ...   
7                     0.026451                        0.023233  ...   
8                     0.063925                        0.028673  ...   
9                     0.042807                        0.038594  ...   

   change_kfi_hisp_pooled_p100_se  change_kii_hisp_pooled_p100_se  \
0                      22654.7810                      17733.0140   
1                       7782.7695                       6791.0166   
2                             NaN                             NaN   
3                             NaN                             NaN   
4                      17989.9490                      25904.1700   
5                             N

# Part 2: Exploratory Visual Analysis
## Phase 1: shape of the data

### Variables
[race] -> pode ser todas (pooled), branca (white), preta (black), hispânica (hisp), asiatica (asian) e indígena americano e nativo do Alaska (aian - American Indian and Alaskan Native)
[gender] -> pode ser ambos (pooled), masculino (male) ou feminino (female)
[pctile] -> pode ser 1, 25, 50, 75, 100
- change\_[outcome]\_[race]\_[gender]\_p[pctile] - num - Difference between [outcome]\_[race]\_[gender]\_p[pctile]\_1992 and [outcome]\_[race]\_[gender]\_p[pctile]\_1978
- change\_[outcome]\_[race]\_[gender]\_p[pctile]_se - num - Standard error for the previous variable

### Problems

In [7]:
# Removing standard error columns
se_columns = filtered_df.filter(like="se", axis=1).columns
no_se_filtered_df = filtered_df.drop(columns=se_columns)
print(30*'=')
display(no_se_filtered_df.head(10))

# Counting columns
num_columns = len(no_se_filtered_df.columns)
print(30*'=')
print(f"Number of columns: {num_columns}")

# Counting null values in columns
print(30*'=')
print(f"Number of rows: {len(no_se_filtered_df.index)}")
null_counts = no_se_filtered_df.isna().sum()
null_counts

# Getting average count of null values
avg_null_counts = null_counts.mean()
print(30*'=')
print(f"average null count: {avg_null_counts}")

# Getting columns columns with above the average null counts
null_counts[null_counts > avg_null_counts]

change_kfr_aian_female_p1  change_kfr_asian_female_p1  \
0                        NaN                         NaN   
1                        NaN                         NaN   
2                        NaN                         NaN   
3                        NaN                         NaN   
4                        NaN                         NaN   
5                        NaN                         NaN   
6                        NaN                         NaN   
7                        NaN                         NaN   
8                        NaN                         NaN   
9                        NaN                         NaN   

   change_kfr_black_female_p1  change_kfr_hisp_female_p1  \
0                    0.007220                        NaN   
1                    0.049207                  -0.057139   
2                    0.007344                        NaN   
3                    0.008007                        NaN   
4                         NaN                  -0.034755   
5                    0.046668                        NaN   
6                    0.010766                        NaN   
7                    0.093872                   0.184147   
8                    0.033095                        NaN   
9                         NaN                        NaN   

   change_kfr_pooled_female_p1  change_kfr_white_female_p1  \
0                    -0.060184                   -0.092919   
1                     0.020877                    0.008520   
2                     0.004866                   -0.052370   
3                    -0.043698                   -0.064093   
4                    -0.032124                   -0.030153   
5                     0.035805                    0.395952   
6                     0.055381                    0.124251   
7                     0.026451                   -0.017449   
8                     0.063925                    0.095082   
9                     0.042807                    0.042807   

   change_kfr_aian_male_p1  change_kfr_asian_male_p1  \
0                      NaN                       NaN   
1                      NaN                       NaN   
2                      NaN                       NaN   
3                      NaN                       NaN   
4                      NaN                       NaN   
5                      NaN                       NaN   
6                      NaN                       NaN   
7                      NaN                       NaN   
8                      NaN                       NaN   
9                      NaN                       NaN   

   change_kfr_black_male_p1  change_kfr_hisp_male_p1  ...  \
0                  0.068086                -0.849591  ...   
1                  0.073520                 0.396834  ...   
2                  0.087051                      NaN  ...   
3                  0.121627                      NaN  ...   
4                       NaN                 0.244030  ...   
5                  0.021109                      NaN  ...   
6                  0.052574                      NaN  ...   
7                  0.052671                 0.003514  ...   
8                  0.004946                      NaN  ...   
9                       NaN                      NaN  ...   

   change_kfi_asian_pooled_p100  change_kii_asian_pooled_p100  \
0                           NaN                           NaN   
1                           NaN                           NaN   
2                           NaN                           NaN   
3                           NaN                           NaN   
4                           NaN                           NaN   
5                           NaN                           NaN   
6                           NaN                           NaN   
7                     -57087.27                    -59150.531   
8                           NaN                           NaN   
9                           NaN                           N

Number of columns: 540
Number of rows: 3191
average null count: 1588.7740740740742


change_kfr_aian_female_p1       2910
change_kfr_asian_female_p1      2733
change_kfr_black_female_p1      1945
change_kfr_hisp_female_p1       1943
change_kfr_aian_male_p1         2895
                                ... 
change_kii_asian_pooled_p100    2545
change_kfi_black_pooled_p100    1759
change_kii_black_pooled_p100    1766
change_kfi_hisp_pooled_p100     1710
change_kii_hisp_pooled_p100     1721
Length: 340, dtype: int64

São 340 (muito mais que a metade) colunas com o número de valores nulos acima da média. Isso significa que a maioria das linhas nessas colunas são nulas. Isso provavelmente está relacionado à estados ou regiões que não tem tantos dados.

In [20]:
# Getting table with population for each state
tables = pd.read_html("https://countryeconomy.com/demography/population/usa-states?year=1978", converters={'CIK': str})
df_states_pop = tables[0].head()
df_states_pop.head(10)
df_states_ordered = df_states_pop.sort_values(by='Population', ascending=False)[["States", "Population"]]

# Renaming columns
df_states_ordered.columns = df_states_ordered.columns.str.lower()
df_states_ordered = df_states_ordered.rename(columns={"states": "state"})

# Correcting values in the states column
df_states_ordered['states'] = df_states_ordered['states'].str.replace(" [+]", "")
df_states_ordered

states  population
4  California    22838960
1     Alabama     3834120
3     Arizona     2517852
2    Arkansas     2241019
0      Alaska      402191

In [9]:
# Colocando de volta coluna de estados
no_se_filtered_df['state'] = df_county_trends_estimate['state_name']
no_se_filtered_df['state'].unique()

array(['Alabama', 'Alaska', 'Arizona', 'Arkansas', 'California',
       'Colorado', 'Connecticut', 'Delaware', 'District of Columbia',
       'Florida', 'Georgia', 'Hawaii', 'Idaho', 'Illinois', 'Indiana',
       'Iowa', 'Kansas', 'Kentucky', 'Louisiana', 'Maine', 'Maryland',
       'Massachusetts', 'Michigan', 'Minnesota', 'Mississippi',
       'Missouri', 'Montana', 'Nebraska', 'Nevada', 'New Hampshire',
       'New Jersey', 'New Mexico', 'New York', 'North Carolina',
       'North Dakota', 'Ohio', 'Oklahoma', 'Oregon', 'Pennsylvania',
       'Rhode Island', 'South Carolina', 'South Dakota', 'Tennessee',
       'Texas', 'Utah', 'Vermont', 'Virginia', 'Washington',
       'West Virginia', 'Wisconsin', 'Wyoming', 'Puerto Rico'],
      dtype=object)